crates/pico_macros/src/memo.rs

```diff
@@ -124,27 +132,90 @@ pub(crate) fn memo(args: TokenStream, item: TokenStream) -> TokenStream {
                 ReturnType::Type(parse_quote!(->), Box::new(parse_quote!(&'db #return_type)));
         }
     } else {
-        return_expr = quote! { #return_expr.clone() };
+        new_sig.generics.params.push(parse_quote!('db));
+        if let FnArg::Typed(PatType { ty, .. }) = &mut new_sig.inputs[0] {
+            if let syn::Type::Reference(ref mut reference) = **ty {
+                reference.lifetime = Some(parse_quote!('db));
+            } else {
+                return Error::new_spanned(ty, "Expected a mutable reference type")
+                    .to_compile_error()
+                    .into();
+            }
+        }
+        new_sig.output = ReturnType::Type(
+            parse_quote!(->),
+            Box::new(parse_quote!(::pico::MemoRef<'db, #return_type>)),
+        );
+        return_expr = quote! {
+            ::pico::MemoRef::new(#db_arg, derived_node_id)
```

Comment:
so the memoized function now returns a `MemoRef<T>` which the user can deref into the actual value?

What's the advantage of doing that, instead of deref'ing for the user? Is it to encourage the use of `MemoRef<T>` as a parameter to memoized functions, so that we can take advantage of the trick above and below?

Reply:
Exactly. We need a way to recognise that an argument is already in the database